blockly-rtc/src/MarkerManager.js

```diff
@@ -0,0 +1,154 @@
+    this.sendMarkerUpdate = sendMarkerUpdate;
+    this.getMarkerUpdates = getMarkerUpdates;
+    this.broadcastMarkerHandler = getBroadcastMarkerUpdates;
+  };
+
+  /**
+   * Initialize the workspace by creating and registering markers for all active
+   * users and activating the handling of recieving MarkerUpdates from the server.
+   * @public
+   */
+  async initMarkers() {
+    const markerUpdates = await this.getMarkerUpdates();
+    markerUpdates.forEach((markerUpdate) => {
+      this.createMarker_(markerUpdate);
+    });
+    if (this.broadcastMarkerHandler) {
+      this.broadcastMarkerHandler(this.updateMarkerLocations_.bind(this));
+    } else {
+      this.pollServer_();
+    };
+};
+
+  /**
+   * Create a MarkerUpdate from a Blockly event and send it to the server.
+   * @param {!Blockly.Event} event The event for which to create a MarkerUpdate.
+   * @public
+   */
+  async handleEvent(event) {
+    const markerUpdate = MarkerUpdate.fromEvent(event);
+    await this.sendMarkerUpdate(markerUpdate);
+  };
+
+  /**
+   * Periodically query the database for MarkerUpdates.
+   * @private
+   */
+  async pollServer_() {
+    const markerUpdates = await this.getMarkerUpdates();
+    await this.updateMarkerLocations_(markerUpdates);
+    setTimeout(() => {
+      this.pollServer_();
+    }, 5000)
+  };
+
+  /**
+   * Get the workspace with the id specified by the MarkerUpdate.
+   * @return {Blockly.Workspace} The sought after workspace or null if not found.
+   * @private
+   */  
+  getWorkspace_() {
+    return Blockly.Workspace.getById(this.workspaceId);
+  };
+
+  /**
+   * Get the MarkerManager for the workspace specified by MarkerUpdate.
+   * @return {Blockly.MarkerManager} The sought after MarkerManager or null if not
+   * found.
+   * @private
+   */  
+  getMarkerManager_() {
+    return this.getWorkspace_() ?
+        this.getWorkspace_().getMarkerManager(): null;
+  };
+
+  /**
+   * Get a color to assign to a new client marker.
+   * @returns {string} The string encoding a colour.
+   * @private
+   */  
+  getColour_() {
+    const colour = this.colours.shift();
+    this.colours.push(colour);
+    return colour;
+  };
+
+  /**
+   * Create a Marker with a unique color and register it.
+   * @param {!MarkerUpdate} markerUpdate The MarkerUpdate for which to create
+   * a Marker.
+   * @returns {!Marker} marker The Marker represented by the MarkerUpdate.
+   * @private
+   */
+  createMarker_(markerUpdate) {
+    if (!this.getMarkerManager_()) {
+      return;
+    };
+    const marker = markerUpdate.toMarker(this.getWorkspace_());
+    marker.colour = this.getColour_();
+    this.getMarkerManager_().registerMarker(markerUpdate.id, marker)
+    marker.setCurNode(markerUpdate.createNode(this.getWorkspace_()));
+    return marker;
+  };
+
+  /**
+   * Get the Marker that corresponds to the given client.
+   * @param {!string} workspaceId The workspaceId of the client.
```

Comment:
type ``{string}`` is ``{!string}`` by default. (ie: not null). The ``!`` is redundant.

Reply:
Done.